src/ultk/language/grammar.py

```diff
@@ -171,6 +171,15 @@ def to_dict(self) -> dict:
         if self.children:
             the_dict["children"] = tuple(child.to_dict() for child in self.children)
         return the_dict
+    
+    # Following function counts the total number of atoms / leaf nodes, as opposed to __len__, which counts all nodes
+    def count_atoms(self):
+        if self.children is None:
+            return 1
+        length = 0
+        if self.children is not None:
+            length += sum(child.count_atoms() for child in self.children)
+        return length
```

Comment:
Nice!  This can be streamlined slightly: after the original `if` statement on line 177, can just do `return sum(child.count_atoms() for child in self.children)` :)

Reply:
Oops! I copied the code over from `__len__` without realizing that it was a `if` statement, not a `for` loop, I have pushed a correction. Thank you for pointing it out.